File: sdk-typescript/src/orchestrator.ts
/**
 * High-level orchestration for Ask-a-Human.
 *
 * This module provides the AskHumanOrchestrator class which handles polling,
 * timeouts, and exponential backoff for async human-in-the-loop workflows.
 *
 * @example
 * ```typescript
 * import { AskHumanClient, AskHumanOrchestrator } from "@ask-a-human/sdk";
 *
 * const client = new AskHumanClient({ agentId: "my-agent" });
 * const orchestrator = new AskHumanOrchestrator(client);
 *
 * const submission = await orchestrator.submit({ prompt: "...", type: "text" });
 * const responses = await orchestrator.awaitResponses([submission.questionId]);
 * ```
 */

import type { AskHumanClient } from "./client.js";
import { AbortError, TimeoutError } from "./errors.js";
import type {
  AwaitResponsesOptions,
  OrchestratorOptions,
  QuestionResponse,
  QuestionSubmission,
  SubmitQuestionOptions,
} from "./types.js";

/** Default poll interval in milliseconds (30 seconds) */
const DEFAULT_POLL_INTERVAL = 30000;

/** Default maximum backoff in milliseconds (5 minutes) */
const DEFAULT_MAX_BACKOFF = 300000;

/** Default backoff multiplier */
const DEFAULT_BACKOFF_MULTIPLIER = 1.5;

/** Default timeout in milliseconds (1 hour) */
const DEFAULT_TIMEOUT = 3600000;

/**
 * High-level orchestrator for async human-in-the-loop workflows.
 *
 * The orchestrator handles common patterns like polling for responses,
 * waiting with timeouts, and exponential backoff. It wraps an AskHumanClient
 * and provides higher-level methods.
 *
 * @example
 * ```typescript
 * const client = new AskHumanClient({ agentId: "my-agent" });
 * const orchestrator = new AskHumanOrchestrator(client, { pollInterval: 30000 });
 *
 * // Submit and wait for responses
 * const submission = await orchestrator.submit({
 *   prompt: "What tone should this use?",
 *   type: "multiple_choice",
 *   options: ["Formal", "Casual"]
 * });
 *
 * const responses = await orchestrator.awaitResponses(
 *   [submission.questionId],
 *   { minResponses: 5, timeout: 300000 }
 * );
 * ```
 */
export class AskHumanOrchestrator {
  /** The underlying AskHumanClient */
  readonly client: AskHumanClient;

  /** Base interval between polls in milliseconds */
  readonly pollInterval: number;

  /** Maximum interval between polls in milliseconds */
  readonly maxBackoff: number;

  /** Multiplier for exponential backoff */
  readonly backoffMultiplier: number;

  /**
   * Initialize the orchestrator.
   *
   * @param client - The AskHumanClient to use for API calls
   * @param options - Orchestrator configuration options
   * @param options.pollInterval - Base interval between polls in milliseconds. Defaults to 30000 (30 seconds)
   * @param options.maxBackoff - Maximum interval between polls in milliseconds. Defaults to 300000 (5 minutes)
   * @param options.backoffMultiplier - Multiplier for exponential backoff. Defaults to 1.5
   *
   * @example
   * ```typescript
   * const client = new AskHumanClient({ agentId: "my-agent" });
   * const orchestrator = new AskHumanOrchestrator(client, {
   *   pollInterval: 15000,  // Start with 15s polls
   *   maxBackoff: 120000,   // Max 2 minutes between polls
   * });
   * ```
   */
  constructor(client: AskHumanClient, options?: OrchestratorOptions) {
    this.client = client;
    this.pollInterval = options?.pollInterval ?? DEFAULT_POLL_INTERVAL;
    this.maxBackoff = options?.maxBackoff ?? DEFAULT_MAX_BACKOFF;
    this.backoffMultiplier = options?.backoffMultiplier ?? DEFAULT_BACKOFF_MULTIPLIER;
  }

  /**
   * Submit a question for humans to answer.
   *
   * This is a convenience wrapper around client.submitQuestion().
   *
   * @param options - Question submission options
   * @returns QuestionSubmission with questionId, status, pollUrl, and expiresAt
   *
   * @example
   * ```typescript
   * const submission = await orchestrator.submit({
   *   prompt: "Which headline is better?",
   *   type: "multiple_choice",
   *   options: ["Option A", "Option B"],
   *   minResponses: 10
   * });
   * console.log(submission.questionId);
   * ```
   */
  async submit(options: SubmitQuestionOptions): Promise<QuestionSubmission> {
    return this.client.submitQuestion(options);
  }

  /**
   * Poll for the current status of questions.
   *
   * This is a non-blocking call that returns immediately with the current
   * status of each question.
   *
   * @param questionIds - List of question IDs to check
   * @returns Record mapping questionId to QuestionResponse
   *
   * @example
   * ```typescript
   * const responses = await orchestrator.pollOnce(["q_abc123", "q_def456"]);
   * for (const [qid, response] of Object.entries(responses)) {
   *   console.log(`${qid}: ${response.status} (${response.currentResponses} responses)`);
   * }
   * ```
   */
  async pollOnce(questionIds: string[]): Promise<Record<string, QuestionResponse>> {
    const results: Record<string, QuestionResponse> = {};

    // Poll all questions in parallel
    const promises = questionIds.map(async (questionId) => {
      const response = await this.client.getQuestion(questionId);
      results[questionId] = response;
    });

    await Promise.all(promises);

    return results;
  }

  /**
   * Wait for responses to questions.
   *
   * Polls periodically until either:
   * - All questions have at least minResponses, or
   * - All questions are CLOSED or EXPIRED, or
   * - The timeout is reached, or
   * - The operation is aborted via AbortSignal
   *
   * Uses exponential backoff to reduce API load during long waits.
   *
   * @param questionIds - List of question IDs to wait for
   * @param options - Options for waiting
   * @param options.minResponses - Minimum responses needed per question. Defaults to 1
   * @param options.timeout - Maximum time to wait in milliseconds. Defaults to 3600000 (1 hour)
   * @param options.signal - AbortSignal for cancellation
   *
   * @returns Record mapping questionId to QuestionResponse. May return partial results if timeout is reached.
   *
   * @throws {TimeoutError} If timeout is reached (only when no partial results available)
   * @throws {AbortError} If the operation is aborted via AbortSignal
   *
   * @example
   * ```typescript
   * // Wait up to 5 minutes for at least 5 responses
   * const responses = await orchestrator.awaitResponses(
   *   ["q_abc123"],
   *   { minResponses: 5, timeout: 300000 }
   * );
   *
   * const question = responses["q_abc123"];
   * if (question.status === "CLOSED") {
   *   console.log("Got all responses!");
   * } else if (question.status === "PARTIAL") {
   *   console.log(`Got ${question.currentResponses} partial responses`);
   * }
   * ```
   */
  async awaitResponses(
    questionIds: string[],
    options?: AwaitResponsesOptions
  ): Promise<Record<string, QuestionResponse>> {
    const minResponses = options?.minResponses ?? 1;
    const timeout = options?.timeout ?? DEFAULT_TIMEOUT;
    const signal = options?.signal;

    const startTime = Date.now();
    let currentInterval = this.pollInterval;
    let results: Record<string, QuestionResponse> = {};

    while (true) {
      // Check for abort
      if (signal?.aborted) {
        throw new AbortError("Operation aborted");
      }

      // Poll all questions
      results = await this.pollOnce(questionIds);

      // Check if all questions are done
      const allDone = this.checkAllDone(results, minResponses);

      if (allDone) {
        return results;
      }

      // Check timeout
      const elapsed = Date.now() - startTime;
      if (elapsed >= timeout) {
        // Return partial results
        return results;
      }

      // Calculate sleep time (don't sleep past timeout)
      const remaining = timeout - elapsed;
      const sleepTime = Math.min(currentInterval, remaining);

      if (sleepTime > 0) {
        await this.sleep(sleepTime, signal);
      }

      // Exponential backoff for next iteration
      currentInterval = Math.min(currentInterval * this.backoffMultiplier, this.maxBackoff);
    }
  }

  /**
   * Submit a question and wait for responses.
   *
   * Convenience method that combines submit() and awaitResponses().
   *
   * @param options - Question submission options
   * @param awaitOptions - Options for waiting
   *
   * @returns QuestionResponse with status and responses
   *
   * @example
   * ```typescript
   * const response = await orchestrator.submitAndWait(
   *   {
   *     prompt: "Should we proceed with option A or B?",
   *     type: "multiple_choice",
   *     options: ["Option A", "Option B"],
   *     minResponses: 5,
   *   },
   *   { timeout: 300000 }  // Wait up to 5 minutes
   * );
   *
   * console.log(`Status: ${response.status}`);
   * for (const r of response.responses) {
   *   console.log(`  Option ${r.selectedOption}`);
   * }
   * ```
   */
  async submitAndWait(
    options: SubmitQuestionOptions,
    awaitOptions?: AwaitResponsesOptions
  ): Promise<QuestionResponse> {
    const submission = await this.submit(options);

    // Use the submission's minResponses if not specified in await options
    const minResponses = awaitOptions?.minResponses ?? options.minResponses ?? 1;

    // Use the submission's timeout if not specified in await options
    const timeout =
      awaitOptions?.timeout ?? (options.timeoutSeconds ? options.timeoutSeconds * 1000 : undefined);

    const results = await this.awaitResponses([submission.questionId], {
      ...awaitOptions,
      minResponses,
      timeout,
    });

    const result = results[submission.questionId];
    if (!result) {
      throw new TimeoutError(
        `No response received for question ${submission.questionId}`,
        awaitOptions?.timeout
      );
    }

    return result;
  }

  /**
   * Check if all questions are done.
   *
   * A question is done if:
   * - It has enough responses (>= minResponses)
   * - It's CLOSED
   * - It's EXPIRED
   *
   * @internal
   */
  private checkAllDone(
    results: Record<string, QuestionResponse>,
    minResponses: number
  ): boolean {
    for (const response of Object.values(results)) {
      // Question is done if it's CLOSED or EXPIRED
      if (response.status === "CLOSED" || response.status === "EXPIRED") {
        continue;
      }

      // Question is done if it has enough responses
      if (response.currentResponses >= minResponses) {
        continue;
      }

      // This question is not done
      return false;
    }

    return true;
  }

  /**
   * Sleep for a specified duration, with abort support.
   *
   * @internal
   */
  private sleep(ms: number, signal?: AbortSignal): Promise<void> {
    return new Promise((resolve, reject) => {
      if (signal?.aborted) {
        reject(new AbortError("Operation aborted"));
        return;
      }

      const timeoutId = setTimeout(resolve, ms);

      if (signal) {
        const abortHandler = () => {
          clearTimeout(timeoutId);
          reject(new AbortError("Operation aborted"));
        };

        signal.addEventListener("abort", abortHandler, { once: true });

        // Clean up the abort handler when the timeout completes
        const originalResolve = resolve;
        resolve = () => {
          signal.removeEventListener("abort", abortHandler);
          originalResolve();
        };
      }
    });
  }
}

File: sdk-python/src/ask_a_human/orchestrator.py
"""High-level orchestration for Ask-a-Human.

This module provides the AskHumanOrchestrator class which handles polling,
timeouts, and exponential backoff for async human-in-the-loop workflows.

Example:
    >>> from ask_a_human import AskHumanClient, AskHumanOrchestrator
    >>> client = AskHumanClient(agent_id="my-agent")
    >>> orchestrator = AskHumanOrchestrator(client)
    >>> submission = orchestrator.submit(prompt="...", type="text")
    >>> responses = orchestrator.await_responses([submission.question_id])
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING

from ask_a_human.types import QuestionResponse, QuestionSubmission

if TYPE_CHECKING:
    from ask_a_human.client import AskHumanClient
    from ask_a_human.types import AudienceTag, QuestionType


# Default configuration
DEFAULT_POLL_INTERVAL = 30.0  # seconds
DEFAULT_TIMEOUT = 3600.0  # 1 hour
DEFAULT_MAX_BACKOFF = 300.0  # 5 minutes max between polls
DEFAULT_BACKOFF_MULTIPLIER = 1.5


class AskHumanOrchestrator:
    """High-level orchestrator for async human-in-the-loop workflows.

    The orchestrator handles common patterns like polling for responses,
    waiting with timeouts, and exponential backoff. It wraps an AskHumanClient
    and provides higher-level methods.

    Attributes:
        client: The underlying AskHumanClient.
        poll_interval: Base interval between polls in seconds.
        max_backoff: Maximum interval between polls in seconds.
        backoff_multiplier: Multiplier for exponential backoff.

    Example:
        >>> client = AskHumanClient(agent_id="my-agent")
        >>> orchestrator = AskHumanOrchestrator(client, poll_interval=30.0)
        >>> # Submit and wait for responses
        >>> submission = orchestrator.submit(
        ...     prompt="What tone should this use?",
        ...     type="multiple_choice",
        ...     options=["Formal", "Casual"]
        ... )
        >>> responses = orchestrator.await_responses(
        ...     [submission.question_id],
        ...     min_responses=5,
        ...     timeout=300
        ... )
    """

    def __init__(
        self,
        client: AskHumanClient,
        poll_interval: float = DEFAULT_POLL_INTERVAL,
        max_backoff: float = DEFAULT_MAX_BACKOFF,
        backoff_multiplier: float = DEFAULT_BACKOFF_MULTIPLIER,
    ) -> None:
        """Initialize the orchestrator.

        Args:
            client: The AskHumanClient to use for API calls.
            poll_interval: Base interval between polls in seconds. Defaults to 30.
            max_backoff: Maximum interval between polls in seconds. Defaults to 300.
            backoff_multiplier: Multiplier for exponential backoff. Defaults to 1.5.

        Example:
            >>> client = AskHumanClient(agent_id="my-agent")
            >>> orchestrator = AskHumanOrchestrator(
            ...     client,
            ...     poll_interval=15.0,  # Start with 15s polls
            ...     max_backoff=120.0,   # Max 2 minutes between polls
            ... )
        """
        self.client = client
        self.poll_interval = poll_interval
        self.max_backoff = max_backoff
        self.backoff_multiplier = backoff_multiplier

    def submit(
        self,
        prompt: str,
        type: QuestionType,  # noqa: A002
        options: list[str] | None = None,
        audience: list[AudienceTag] | None = None,
        min_responses: int = 5,
        timeout_seconds: int = 3600,
        idempotency_key: str | None = None,
    ) -> QuestionSubmission:
        """Submit a question for humans to answer.

        This is a convenience wrapper around client.submit_question().

        Args:
            prompt: The question text (10-2000 characters).
            type: Question type - 'text' for free-form, 'multiple_choice' for options.
            options: Options for multiple choice questions (2-10 items).
            audience: Target audience tags. Defaults to ['general'].
            min_responses: Minimum responses needed (1-50). Defaults to 5.
            timeout_seconds: Question expiration in seconds (60-86400). Defaults to 3600.
            idempotency_key: Optional key to prevent duplicate submissions.

        Returns:
            QuestionSubmission with question_id, status, poll_url, and expires_at.

        Example:
            >>> submission = orchestrator.submit(
            ...     prompt="Which headline is better?",
            ...     type="multiple_choice",
            ...     options=["Option A", "Option B"],
            ...     min_responses=10
            ... )
            >>> print(submission.question_id)
        """
        return self.client.submit_question(
            prompt=prompt,
            type=type,
            options=options,
            audience=audience,
            min_responses=min_responses,
            timeout_seconds=timeout_seconds,
            idempotency_key=idempotency_key,
        )

    def poll_once(self, question_ids: list[str]) -> dict[str, QuestionResponse]:
        """Poll for the current status of questions.

        This is a non-blocking call that returns immediately with the current
        status of each question.

        Args:
            question_ids: List of question IDs to check.

        Returns:
            Dictionary mapping question_id to QuestionResponse.

        Example:
            >>> responses = orchestrator.poll_once(["q_abc123", "q_def456"])
            >>> for qid, response in responses.items():
            ...     print(f"{qid}: {response.status} ({response.current_responses} responses)")
        """
        results: dict[str, QuestionResponse] = {}
        for question_id in question_ids:
            results[question_id] = self.client.get_question(question_id)
        return results

    def await_responses(
        self,
        question_ids: list[str],
        min_responses: int = 1,
        timeout: float = DEFAULT_TIMEOUT,
    ) -> dict[str, QuestionResponse]:
        """Wait for responses to questions.

        Polls periodically until either:
        - All questions have at least min_responses, or
        - All questions are CLOSED or EXPIRED, or
        - The timeout is reached

        Uses exponential backoff to reduce API load during long waits.

        Args:
            question_ids: List of question IDs to wait for.
            min_responses: Minimum responses needed per question. Defaults to 1.
            timeout: Maximum time to wait in seconds. Defaults to 3600 (1 hour).

        Returns:
            Dictionary mapping question_id to QuestionResponse.
            May return partial results if timeout is reached.

        Example:
            >>> # Wait up to 5 minutes for at least 5 responses
            >>> responses = orchestrator.await_responses(
            ...     ["q_abc123"],
            ...     min_responses=5,
            ...     timeout=300
            ... )
            >>> question = responses["q_abc123"]
            >>> if question.status == "CLOSED":
            ...     print("Got all responses!")
            ... elif question.status == "PARTIAL":
            ...     print(f"Got {question.current_responses} partial responses")
        """
        start_time = time.time()
        current_interval = self.poll_interval
        results: dict[str, QuestionResponse] = {}

        while True:
            # Poll all questions
            results = self.poll_once(question_ids)

            # Check if all questions are done
            all_done = True
            for response in results.values():
                # Question is done if:
                # - It has enough responses
                # - It's CLOSED
                # - It's EXPIRED
                if response.status in ("CLOSED", "EXPIRED"):
                    continue
                if response.current_responses >= min_responses:
                    continue
                all_done = False
                break

            if all_done:
                return results

            # Check timeout
            elapsed = time.time() - start_time
            if elapsed >= timeout:
                # Return partial results
                return results

            # Calculate sleep time (don't sleep past timeout)
            remaining = timeout - elapsed
            sleep_time = min(current_interval, remaining)

            if sleep_time > 0:
                time.sleep(sleep_time)

            # Exponential backoff for next iteration
            current_interval = min(
                current_interval * self.backoff_multiplier,
                self.max_backoff,
            )

    def submit_and_wait(
        self,
        prompt: str,
        type: QuestionType,  # noqa: A002
        options: list[str] | None = None,
        audience: list[AudienceTag] | None = None,
        min_responses: int = 5,
        timeout_seconds: int = 3600,
        idempotency_key: str | None = None,
        wait_timeout: float | None = None,
    ) -> QuestionResponse:
        """Submit a question and wait for responses.

        Convenience method that combines submit() and await_responses().

        Args:
            prompt: The question text (10-2000 characters).
            type: Question type - 'text' for free-form, 'multiple_choice' for options.
            options: Options for multiple choice questions (2-10 items).
            audience: Target audience tags. Defaults to ['general'].
            min_responses: Minimum responses needed (1-50). Defaults to 5.
            timeout_seconds: Question expiration in seconds (60-86400). Defaults to 3600.
            idempotency_key: Optional key to prevent duplicate submissions.
            wait_timeout: How long to wait for responses. Defaults to timeout_seconds.

        Returns:
            QuestionResponse with status and responses.

        Example:
            >>> response = orchestrator.submit_and_wait(
            ...     prompt="Should we proceed with option A or B?",
            ...     type="multiple_choice",
            ...     options=["Option A", "Option B"],
            ...     min_responses=5,
            ...     wait_timeout=300  # Wait up to 5 minutes
            ... )
            >>> print(f"Status: {response.status}")
            >>> for r in response.responses:
            ...     print(f"  {response.options[r.selected_option]}")
        """
        submission = self.submit(
            prompt=prompt,
            type=type,
            options=options,
            audience=audience,
            min_responses=min_responses,
            timeout_seconds=timeout_seconds,
            idempotency_key=idempotency_key,
        )

        # Wait for responses
        wait_time = wait_timeout if wait_timeout is not None else float(timeout_seconds)
        results = self.await_responses(
            question_ids=[submission.question_id],
            min_responses=min_responses,
            timeout=wait_time,
        )

        return results[submission.question_id]
